agent/app/agent.go

```diff
@@ -232,6 +232,8 @@ func (agent *ecsAgent) newTaskEngine(containerChangeEventStream *eventstream.Eve
 		log.Warnf(instanceIDMismatchErrorFormat,
 			previousEC2InstanceID, currentEC2InstanceID)
 
+		// Reset agent state as a new container instance
+		state.Reset()
```

Comment:
It seems like you'd need to invoke `statemanager.Save()` here as well? Otherwise, an agent restart here may still result in the wrong state, yes?